endorsedLibraries/kotlinx.cli/src/main/kotlin/kotlinx/cli/ArgParser.kt

```diff
@@ -0,0 +1,467 @@
+
+internal expect fun exitProcess(status: Int): Nothing
+
+/**
+ * Queue of arguments descriptors.
+ * Arguments can have several values, so one descriptor can be returned several times.
+ */
+internal class ArgumentsQueue(argumentsDescriptors: List<ArgDescriptor<*, *>>) {
+    /**
+     * Map of arguments descriptors and their current usage number.
+     */
+    private val argumentsUsageNumber = linkedMapOf(*argumentsDescriptors.map { it to 0 }.toTypedArray())
+
+    /**
+     * Get next descriptor from queue.
+     */
+    fun pop(): String? {
+        if (argumentsUsageNumber.isEmpty())
+            return null
+
+        val (currentDescriptor, usageNumber) = argumentsUsageNumber.iterator().next()
+        currentDescriptor.number?.let {
+            // Parse all arguments for current argument description.
+            if (usageNumber + 1 >= currentDescriptor.number) {
+                // All needed arguments were provided.
+                argumentsUsageNumber.remove(currentDescriptor)
+            } else {
+                argumentsUsageNumber[currentDescriptor] = usageNumber + 1
+            }
+        }
+        return currentDescriptor.fullName
+    }
+}
+
+/**
+ * Interface of argument value.
+ */
+interface ArgumentValueDelegate<T> {
+    val value: T
+    operator fun getValue(thisRef: Any?, property: KProperty<*>): T = value
+    operator fun setValue(thisRef: Any?, property: KProperty<*>, value: T)
+}
+
+/**
+ * Abstract base class for subcommands.
+ */
+@SinceKotlin("1.3")
+@ExperimentalCli
+abstract class Subcommand(val name: String): ArgParser(name) {
+    /**
+     * Execute action if subcommand was provided.
+     */
+    abstract fun execute()
+}
+
+/**
+ * Argument parsing result.
+ * Contains name of subcommand which was called.
+ *
+ * @property commandName name of command which was called.
+ */
+class ArgParserResult(val commandName: String)
+
+/**
+ * Arguments parser.
+ *
+ * @property programName name of current program.
+ * @property useDefaultHelpShortName add or not -h flag for help message.
+ * @property prefixStyle style of expected options prefix.
+ * @property skipExtraArguments just skip extra arguments in command line string without producing error message.
+ */
+open class ArgParser(val programName: String, var useDefaultHelpShortName: Boolean = true,
+                     var prefixStyle: OPTION_PREFIX_STYLE = OPTION_PREFIX_STYLE.LINUX,
+                     var skipExtraArguments: Boolean = false) {
+
+    /**
+     * Map of options: key - full name of option, value - pair of descriptor and parsed values.
+     */
+    private val options = mutableMapOf<String, ParsingValue<*, *>>()
+    /**
+     * Map of arguments: key - full name of argument, value - pair of descriptor and parsed values.
+     */
+    private val arguments = mutableMapOf<String, ParsingValue<*, *>>()
+
+    /**
+     * Map with declared options.
+     */
+    private val declaredOptions = mutableMapOf<Int, OptionImpl<*, *>>()
```

Comment:
Need such maps, because modifiers can change option several times, before it's named.